tmt/steps/execute/__init__.py

```diff
@@ -55,77 +58,199 @@
 # Metadata file with details about the current test
 TEST_METADATA_FILENAME = 'metadata.yaml'
 
-# Scripts source directory
+#: Scripts source directory
 SCRIPTS_SRC_DIR = tmt.utils.resource_files('steps/execute/scripts')
 
+#: The default scripts destination directory
+SCRIPTS_DEST_DIR = Path("/usr/local/bin")
+
+#: The default scripts destination directory for rpm-ostree based distributions, https://github.com/teemtee/tmt/discussions/3260
+SCRIPTS_DEST_DIR_OSTREE = Path("/var/lib/tmt/scripts")
+
+#: The default tmt environment variable name for forcing ``SCRIPTS_DEST_DIR``
+SCRIPTS_DEST_DIR_VARIABLE = 'TMT_SCRIPTS_DEST_DIR'
+
 
 @dataclass
 class Script:
-    """ Represents a script provided by the internal executor """
+    """
+    Represents a script provided by the internal executor.
+
+    Must be used as a context manager. The context manager returns
+    the source file path.
+
+    The source file is defined by the ``source_filename`` attribute and its
+    location is relative to the directory specified via the :py:data:`SCRIPTS_SRC_DIR`
+    variable. All scripts must be located in this directory.
 
-    path: Path
-    aliases: list[Path]
+    The default destination directory of the scripts is :py:data:`SCRIPTS_DEST_DIR`.
+    On ``rpm-ostree`` distributions like Fedora CoreOS, the default destination
+    directory is :py:data:``SCRIPTS_DEST_DIR_OSTREE``. The destination directory
+    of the scripts can be forced by the script using ``destination_path`` attribute.
+
+    The destination directory can be overridden using the environment variable defined
+    by the :py:data:`SCRIPTS_DEST_DIR_VARIABLE` variable.
+
+    The ``enabled`` attribute can specify a function which is called with :py:class:`Guest`
+    instance to evaluate if the script is enabled. This can be useful to optionally disable
+    a script for specific guests.
+    """
+
+    source_filename: str
+    destination_path: Optional[Path]
+    aliases: list[str]
     related_variables: list[str]
+    enabled: Callable[[Guest], bool]
+
+    def __enter__(self) -> Path:
+        return SCRIPTS_SRC_DIR / self.source_filename
+
+    def __exit__(self, *args: object) -> None:
+        pass
+
+    def guest_path(self, guest: Guest, alias: Optional[str] = None) -> Path:
+        """ Return path of the script or the script alias on a guest """
+        filename = alias or self.source_filename
+        scripts_dest_dir = effective_scripts_dest_dir(
+            default=SCRIPTS_DEST_DIR_OSTREE if guest.facts.is_ostree else SCRIPTS_DEST_DIR
+            )
+        return self.destination_path if self.destination_path else scripts_dest_dir / filename
```

Comment:
* Is `alias` ever optional? In `script.guest_path(guest, dest)`, `dest` - which I proposed to rename to `filename` - comes from `[script.source_filename, *script.aliases]`, `source_filename` is `str, `script.aliases` is `list[str]`, all in all, `alias` will never be `None`. And `source_filename` is one of the inputs, therefore `filename = alias or self.source_filename` is not needed.
* I'd recommend changing `alias` to `filename` or something, sometimes it's alias, sometimes it's `source_filename`.
* if `self.destination_path` is set, you can return the final path immediately, and only after that you can construct `scripts_dest_dir`.
* `prepare_scripts()` already got hold of the actual `scripts_dest_dir`, hm, do we need this method at all? If the script has `destination_path`, `prepare_scripts` can use it, if it's not set, `prepare_scripts` can do `scripts_dest_dir / filename` because it already has `scripts_dest_dir`.

Reply:
I do believe I addressed all, except the last one. I added this to mitigate duplication in:

tmt/frameworks/beakerlib.py

where I needed ti for:

```
   36         return Environment(data: {
   37             'BEAKERLIB_DIR': EnvVarValue(raw_value: invocation.path),
   38             'BEAKERLIB_COMMAND_SUBMIT_LOG': EnvVarValue(
   39                 raw_value: f'bash {tmt.steps.execute.TMT_FILE_SUBMIT_SCRIPT.guest_path(guest: invocation.guest)}')
   40             })
```